### PR TITLE
Feature/hir 122 add endpoint to upload room photos

### PR DESCRIPTION
**Pull request by AlexeyShok**
New model works with images, added images for rooms. Accommodation and room included ordered list of slim images. Upgrated OpenTelemetry version too.

### CODE REVIEW

File: HappyTravel.Hiroshima.Data/DirectContractsDbContext.cs [hunk 1]
```diff
@@ -86,6 +86,7 @@ private void AddAccommodations(ModelBuilder modelBuilder)
                 e.Property(a => a.Modified).IsRequired();
                 e.Property(a => a.Floors);
                 e.Property(a => a.BuildYear);
+                e.Property(a => a.Images).HasColumnType("jsonb").IsRequired().HasDefaultValueSql("'[]'::json");
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.Data/DirectContractsDbContext.cs:89]: maybe `"'[]'::jsonb"` ?

File: HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs [hunk 2]
```diff
@@ -103,6 +103,92 @@ public async Task<IActionResult> RemoveImageFile([FromRoute] int accommodationId
         }
 
 
+        /// <summary>
+        /// Retrieves ordered list of images by accommodation ID and room ID
+        /// </summary>
+        /// <param name="accommodationId">ID of the accommodation</param>
+        /// <param name="roomId">ID of the room</param>
+        /// <returns></returns>
+        [HttpGet("accommodations/{accommodationId}/rooms/{roomId}/photo")]
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:112]: maybe `photos` ?

File: HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs [hunk 3]
```diff
@@ -103,6 +103,92 @@ public async Task<IActionResult> RemoveImageFile([FromRoute] int accommodationId
...
+
+
+        /// <summary>
+        /// Updates the list of images for room by Ids
+        /// </summary>
+        /// <param name="accommodationId">Accommodation Id</param>
+        /// <param name="roomId">Room Id</param>
+        /// <param name="slimImages">Ordered list of images</param>
+        /// <returns></returns>
+        [HttpPut("accommodations/{accommodationId}/rooms/{roomId}/photo")]
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:159]: `photos` ?

File: HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs [hunk 4]
```diff
@@ -103,6 +103,92 @@ public async Task<IActionResult> RemoveImageFile([FromRoute] int accommodationId
...
 
 
+        /// <summary>
+        /// Retrieves ordered list of images by accommodation ID and room ID
+        /// </summary>
+        /// <param name="accommodationId">ID of the accommodation</param>
+        /// <param name="roomId">ID of the room</param>
+        /// <returns></returns>
+        [HttpGet("accommodations/{accommodationId}/rooms/{roomId}/photo")]
+        [ProducesResponseType(typeof(Hiroshima.DirectManager.Models.Responses.Room), (int)HttpStatusCode.OK)]
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:113]: Wrong response type

File: HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs [hunk 5]
```diff
@@ -103,6 +103,92 @@ public async Task<IActionResult> RemoveImageFile([FromRoute] int accommodationId
...
+        /// Deletes image file by ID
+        /// </summary>
+        /// <param name="accommodationId">Accommodation Id</param>
+        /// <param name="roomId">Room Id</param>
+        /// <param name="imageId">Id of the image file to be deleted</param>
+        /// <returns></returns>
+        [HttpDelete("accommodations/{accommodationId}/rooms/{roomId}/photo/{imageId}")]
+        [ProducesResponseType((int)HttpStatusCode.NoContent)]
+        [ProducesResponseType(typeof(ProblemDetails), (int)HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> RemoveImageFile([FromRoute] int accommodationId, [FromRoute] int roomId, [FromRoute] Guid imageId)
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:182]: It's not necessary to use  "File" in the method's name. The same with other methods in this controller.

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 6]
```diff
@@ -42,180 +44,86 @@ public Task<Result<List<Models.Responses.SlimImage>>> Get(int accommodationId)
...
-                    void SetImageDetails()
-                    {
-                        var basePartOfKey = $"{S3FolderName}/{dbImage.AccommodationId}/{imageId}";
-                        dbImage.MainImage.Key = $"{basePartOfKey}-main.jpg";
-                        dbImage.SmallImage.Key = $"{basePartOfKey}-small.jpg";
+        public Task<Result<Guid>> Add(Models.Requests.AccommodationImage image)
+        {
+            return _contractManagerContext.GetContractManager()
+                .EnsureAccommodationBelongsToContractManager(_dbContext, image.AccommodationId)
+                .Tap(contractManager => ValidationHelper.Validate(image, new AccommodationImageValidator()))
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:74]: According to our agreement Tap can't interrupt the chain of methods.

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 7]
```diff
@@ -314,6 +407,142 @@ private async Task ResortImages(int contractManagerId, int accommodationId)
...
+            _dbContext.Rooms.Update(room);
+            await _dbContext.SaveChangesAsync();
+
+            return Result.Success();
+        }
+
+
+        private SlimImage Build(Image image)
+        {
+            return  new SlimImage
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:500]: Extra space

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 8]
```diff
@@ -314,6 +407,142 @@ private async Task ResortImages(int contractManagerId, int accommodationId)
...
+                LargeImageURL = GetImageUrl(image.Keys.MainImage),
+                SmallImageURL = GetImageUrl(image.Keys.SmallImage),
+                Description = image.Description.GetValue<MultiLanguage<string>>()
+            };
+        }
+
+
+        private List<SlimImage> Build(List<Models.Requests.SlimImage> images)
+        {
+            var slimImages = new List<SlimImage>();
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:512]: The size of the list can be set

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 9]
```diff
@@ -314,6 +407,142 @@ private async Task ResortImages(int contractManagerId, int accommodationId)
...
+                };
+                slimImages.Add(slimImage);
+            }
+            return slimImages;
+        }
+
+
+        private List<Models.Responses.SlimImage> Build(List<Image> images)
+        {
+            var slimImages = new List<Models.Responses.SlimImage>();
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:530]: Capacity can be set

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 10]
```diff
@@ -42,180 +44,86 @@ public Task<Result<List<Models.Responses.SlimImage>>> Get(int accommodationId)
...
+                .Tap(contractManager => ResortImages(contractManager.Id, image.AccommodationId, ImageTypes.AccommodationImage))
+                .Map(contractManager => Create(contractManager.Id, image))
+                .Ensure(dbImage => ValidateImageType(image.UploadedFile).Value, "Invalid image file type")
+                .Bind(dbImage => ConvertAndUpload(dbImage, image.UploadedFile))
+                .Bind(AddSlimImageToAccommodation);
+        }
 
-                        dbImage.MainImage.Url = $"{_basePathToAmazon}/{dbImage.MainImage.Key}";
-                        dbImage.SmallImage.Url = $"{_basePathToAmazon}/{dbImage.SmallImage.Key}";
-                    }
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:82]: Extra lines

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 11]
```diff
@@ -42,180 +44,86 @@ public Task<Result<List<Models.Responses.SlimImage>>> Get(int accommodationId)
...
-                }
-            }
-            
-            Result<bool> ValidateImageType(FormFile uploadedFile)
-            {
-                var extension = Path.GetExtension(uploadedFile?.FileName)?.ToLower() ?? string.Empty;
-                return extension == ".jpg" || extension == ".jpeg" || extension == ".png";
-            }
-        }
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:98]: Extra lines

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 12]
```diff
@@ -42,250 +44,337 @@ public Task<Result<List<Models.Responses.SlimImage>>> Get(int accommodationId)
...
-                        .Branch(f => f.ConstrainWithin(ResizedSmallImageMaximumSideSize, ResizedSmallImageMaximumSideSize).EncodeToBytes(new MozJpegEncoder(TargetJpegQuality, true)))
-                        .EncodeToBytes(new MozJpegEncoder(TargetJpegQuality, true))
-                        .Finish().InProcessAsync();
 
-                    imagesSet.SmallImage = GetImage(1);
-                    imagesSet.MainImage = GetImage(2);
+        public async Task<Result> RemoveAll(int contractManagerId, int accommodationId)
+        {
+            var images = _dbContext.Images.Where(image => image.ContractManagerId == contractManagerId && 
+                image.ReferenceId == accommodationId && image.ImageType == ImageTypes.AccommodationImage);
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:153]: ToListAsync() ?

File: HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs [hunk 13]
```diff
@@ -42,250 +44,337 @@ public Task<Result<List<Models.Responses.SlimImage>>> Get(int accommodationId)
...
+        }
 
-                    return imagesSet.MainImage.Any() && imagesSet.SmallImage.Any()
-                        ? Result.Success(imagesSet)
-                        : Result.Failure<ImageSet>("Processing of the images failed");
 
+        public async Task<Result> RemoveAll(int contractManagerId, int accommodationId, int roomId)
+        {
+            var images = _dbContext.Images.Where(image => image.ContractManagerId == contractManagerId &&
+                image.ReferenceId == roomId && image.ImageType == ImageTypes.RoomImage);
```
Comment (oleg1std) [inline HappyTravel.Hiroshima.DirectManager/Services/ImageManagementService.cs:168]: ToListAsync() ?

File: HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs [hunk 14]
```diff
@@ -103,6 +103,92 @@ public async Task<IActionResult> RemoveImageFile([FromRoute] int accommodationId
...
+
+
+        /// <summary>
+        /// Deletes image file by ID
+        /// </summary>
+        /// <param name="accommodationId">Accommodation Id</param>
+        /// <param name="roomId">Room Id</param>
+        /// <param name="imageId">Id of the image file to be deleted</param>
+        /// <returns></returns>
+        [HttpDelete("accommodations/{accommodationId}/rooms/{roomId}/photo/{imageId}")]
```
Comment (hasmikharutyunyan) [inline HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:179]: I thin here also better to use `photos`

Comment (AlexeyShok) [reply HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:179]: No. In this case, we delete only one image from database. In Get we receive some images. In Put we send a list of several images.

Comment (hasmikharutyunyan) [reply HappyTravel.Hiroshima.WebApi/Controllers/DirectManager/ImageManagementController.cs:179]: Yes I understand, but this is similar with `accommodations/{accommodationId}`, you chose from acommodations accommodation with Id `{accommodationId}`, and  about photos also we can say that we from photos choose photo with Id `{imageId}`.

**Review by hasmikharutyunyan — APPROVED** [review]

**Pull request merged**